### PR TITLE
fix(gnovm): correctly update refs count when self-assigning

### LINKED ISSUE #939: Pop/push slice returns old element
### Description

If you pop an element from a slice (via `slice = slice[:len(slice)-1]`) and append a new element afterwards, the last element of the slice will be the popped element and not the new element

(Sorry for the confusing issue name I can't find a better short name)

### Your environment

* macOS 13.3.1
* go version go1.20.4 darwin/arm64
* gno: master
* commit: d684985f9d146c4b6495508adef57a1ea60ba13a

### Steps to reproduce

- Deploy this realm
```go
package bugmin

import (
	"strconv"
	"strings"
)

var slice = []string{"undead-element"}

func Pop() {
	slice = slice[:len(slice)-1]
}

func Push() {
	slice = append(slice, "new-element")
}

func Render(path string) string {
	str := strings.Join(slice, ", ") + "\n\n"
	str += "len: " + strconv.Itoa(len(slice)) + "\n\n"
	return str
}
```

- Call `Pop` then `Push`

### Expected behaviour

The slice is `[]string{"new-element"}` after `Pop` and `Push`

### Actual behaviour

The slice is `[]string{"undead-element"}` after `Pop` and `Push`

### Additional info

If you give me some pointer (namely where the slice access operator `[:]` and `append` are handled), I could try debug this

You can work around this by creating a new slice:
```go
func Pop() {
	// slice = slice[:len(slice)-1] // this is bugged

	newSlice := []string{}
	for i := 0; i < len(slice)-1; i++ {
		newSlice = append(newSlice, slice[i])
	}
	slice = newSlice // this works
}
```

The bug seem to be at storage layer level since this works as expected:
```go
package bugmin

import (
	"strconv"
	"strings"
)

func Render(path string) string {
	slice := []string{"undead-element"}
	slice = slice[:len(slice)-1]
	slice = append(slice, "new-element")

	str := "slice: " + strings.Join(slice, ", ") + "\n\n"
	str += "len(slice): " + strconv.Itoa(len(slice)) + "\n\n"
	return str
}
```

### PR DESCRIPTION
**Pull request by n0izn0iz**
Fixes #939 
See https://github.com/gnolang/gno/pull/960#issuecomment-1636566357 for explanations of the fix
Thanks to @peter7891 for most of the actual fixing

<details><summary>Checklists...</summary>

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [x] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](../.benchmarks/README.md).

## Maintainers Checklist

- [ ] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [ ] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [ ] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change
</details>

### CODE REVIEW

**Comment by peter7891** [comment]
Before merging anything, we need to answer the following questions convincingly, in the following order.

1. Is the bug reproducible?
2. What is broken?
3. How does this change fix it?

I haven't been able to reproduce the bug, which limits me in making a strong case to answering the other 2 questions.
Can you @n0izn0iz provide me with detailed steps you perform to reproduce it?

I've ran the functions you listed in a  regular Go test inside the VM and Gno test contract, it works fine.

**Comment by n0izn0iz** [comment]
I don't want to merge this, like I said I barely understand my changes
Though I did retry the steps outlined in #939 on latest master and I still get `undead-element` in render after `Pop` and `Push`

**Comment by peter7891** [comment]
> @peter7891 I don't want to merge this, like I said I barely understand my changes (and the commit is called "no idea what I'm doing") Though I did retry the steps outlined in #939 on latest master ([02b88c0](https://github.com/gnolang/gno/commit/02b88c009177dde255c0a5dc0d36203308b2f050)) and I still get `undead-element` in render after `Pop` and `Push` We can do peer-programing so I can show you how to reproduce if you want

That would be great. Discord? Petar#4367

**Comment by peter7891** [comment]
> Before merging anything, we need to answer the following questions convincingly, in the following order.
> 
> 1. Is the bug reproducible?
> 2. What is broken?
> 3. How does this change fix it?
> 
> I haven't been able to reproduce the bug, which limits me in making a strong case to answering the other 2 questions. Can you @n0izn0iz provide me with detailed steps you perform to reproduce it?
> 
> I've ran the functions you listed in a regular Go test inside the VM and Gno test contract (locally), it works fine.

I think, i can answer these questions now.

1. I've reproduced the bug locally.
2. the refcount for the slice, when it's being assigned to itself is increasing inside realm.DidUpdate()
[here](https://github.com/gnolang/gno/blame/master/gnovm/pkg/gnolang/realm.go#L155)

With the code of the original bug report, 

```
var slice = []string{"undead-element"}

func Pop() {
	slice = slice[:len(slice)-1]
}
```
After the assignment inside `Pop` gets evaluated, `slice` has a refcount of `2`, for a brief period, which should not be the case. This makes it being treated as NewEscaped and it doesn't get marked as dirty.

3. I think, we should either create a separate function to handle this corner case or only increment and decrement refs when the old and new object are not the same.

@jaekwon What do you think?

**Comment by n0izn0iz** [comment]
I updated the commit with the changes proposed by @peter7891

**Comment by zivkovicmilos** [comment]
Are we able to reproduce this in a unit test at all?

**Comment by n0izn0iz** [comment]
Yeah I managed to have a test using manfred's pattern in #979, will push it here

**Comment by n0izn0iz** [comment]
The test I just added fails on master but passes on this PR, it reproduces the original bug description in #939

File: examples/gno.land/r/demo/multitxtest/multitx_testing.go
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/multitxtest/multitx_testing.go:1]: Should these tests be contained in the `examples` subdirectory, instead of `gnovm`, since these are `.go` files after all?

**Review by zivkovicmilos — APPROVED** [review]
Thank you for adding a fix for this 🙏 
Can you please update the PR description with a brief summary of the changes (from Peter's comment)?

Pinging @peter7891 and @jaekwon to give a look as well

File: examples/gno.land/r/demo/multitxtest/multitx_testing.go
Comment (moul) [reply examples/gno.land/r/demo/multitxtest/multitx_testing.go:1]: But can you merge the two files into a single `_test.go` file instead of creating this API which won't be reused elsewhere, please?

**Review by moul — CHANGES_REQUESTED** [review]
See comments, thank you.

**Comment by zivkovicmilos** [comment]
@n0izn0iz Ping for visibility on this PR, please check out the comments so we can go ahead and merge this 🙏

**Comment by moul** [comment]
@n0izn0iz just switched the PR to draft; please remove the draft flag when you'll apply the last changes.

**Comment by n0izn0iz** [comment]
sorry for moving slowly on that, I'd like to take the time to properly implement the multi-tx test helper, currently it barely works, I barely understand it and is very hacky/brittle, maybe I should revert to include it as-is directly in the test and do the mutli-tx test helper in another PR

**Comment by tbruyelle** [comment]
@n0izn0iz Hey sounds this is related to a bug we had on gnochess, see #1170, can you confirm we're talking about the same bug ?
About the multi tx test helper, you should have a look at the recent work from @moul & @gfanton  with testscript on gnochess : https://github.com/gnolang/gnochess/commit/df9744dc48107cb9515939b698164aba47b2884f

Basically you setup your "*_test.go" file with `gno/gno.land/pkg/integration` (unfortunately not yet merged on master, still in the patch/gnochess branch), and then you create `txtar` files that are executed by your setup. The txtar files contain commands that looks like shell, but some of them are overridden, some other are customs like `cmp`. At the end of the file you can generate files with the `--- filename ---` instruction, those files will be created before the execution of the commands from above.

**Comment by deelawn** [comment]
The issue this PR was meant to address has been fixed by #1305

**Pull request closed without merging**
